csrc/device_lower/pass/index.cpp

```diff
@@ -1742,10 +1743,240 @@ static MmaInputSmemSwizzle getSwizzleMode(TensorView* tv) {
+// stride of the outer (m_inst/8) in the allocation domain of tv.
+// That is, if we futher split m_inst as:
+//       M            K
+//       |            |
+//      ...          ...
+//       |            |
+//      / \.         / \.
+//    ...  m_inst  ...  k_inst
+//         /    \.
+//   linear      8
+// Then we would need to prove that `linear` is linear in the allocation domain
+// of tv, and the stride of `linear` is the stride_bytes. This function does
+// the following things:
+// 1. Find m_inst.
+// 2. Split m_inst as above.
+// 3. Prove that `linear` is linear in the allocation domain of tv, and get the
+//    stride of `linear`.
+Val* getOuterStride(TensorView* tv, const MmaOp* mma) {
+  ValGraph& id_graph = GpuLower::current()->tensorIndexer().traversalGraph();
+  auto logical_domain =
+      id_graph.toGroups(TensorDomain::noBroadcasts(tv->getLogicalDomain()));
+  auto loop_domain =
+      id_graph.toGroups(mma->out()->as<TensorView>()->getLoopDomain());
+  auto alloc_domain = id_graph.toGroups(tv->getMaybeAllocationDomain());
+
+  // In the consumer's loop domain, there should be exactly 3 IDs parallelized
+  // on Mma. Which of these three dims are M, N, and K? We don't know. But we
+  // don't really care. What we do care is, which is the inner? which is the
+  // broadcast? We would find the loop dim that is neither inner nor broadcast,
+  // and the stride of that dim is the one we are looking for.
+  auto inner = getInnerMmaLoopGroup(tv, mma);
+  std::vector<ValGroup> mma_groups;
+  mma_groups.reserve(2);
+  for (auto id : mma->out()->as<TensorView>()->getLoopDomain()) {
+    if (id->getParallelType() == ParallelType::Mma && !inner->has(id)) {
+      mma_groups.emplace_back(id_graph.toGroup(id));
+    }
+  }
+  NVF_ERROR(
+      mma_groups.size() == 2,
+      "Expecting 3 IDs in the loop domain of mma output to be parallelized on Mma,",
+      " among which one must be the innermost of producer's allocation domain");
+
+  // Project mma_groups to the logical domain of tv.
+  std::vector<std::unordered_set<ValGroup>> projections_on_logical;
+  projections_on_logical.reserve(mma_groups.size());
+  for (const auto& g : mma_groups) {
+    projections_on_logical.push_back({g});
+  }
+  auto exprs =
+      ValGraphBFS::getExprsBetween(id_graph, loop_domain, logical_domain);
+  for (const auto& [expr, direction] : exprs) {
+    auto from =
+        (direction == Direction::Forward ? id_graph.inputGroups(expr)
+                                         : id_graph.outputGroups(expr));
+    auto to =
+        (direction == Direction::Forward ? id_graph.outputGroups(expr)
+                                         : id_graph.inputGroups(expr));
+    for (const auto& g : from) {
+      for (auto& projection_on_logical : projections_on_logical) {
+        if (projection_on_logical.count(g)) {
+          projection_on_logical.erase(g);
+          projection_on_logical.insert(to.begin(), to.end());
+        }
+      }
+    }
+  }
```

Comment:
Projecting a subset of a domain onto another domain feels like a utility that would be useful elsewhere. Maybe consider making it a separate utility?

Reply:
changed